core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java

```diff
@@ -0,0 +1,130 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.rest;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * {@code DeprecationRestChannel} is a proxy for an existing {@link RestChannel} that decorates the {@link RestResponse} with
+ * the {@link #DEPRECATION_HEADER deprecation header} and a warning message to denote that the existing request/response is
+ * deprecated.
+ */
+public class DeprecationRestChannel implements RestChannel {
+    /**
+     * The "Warning" Header comes from RFC-7234. As the RFC describes, it's generally used for caching purposes, but it can be
+     * used for <em>any</em> warning.
+     *
+     * https://tools.ietf.org/html/rfc7234#section-5.5
+     */
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * The proxied canchannel.
+     */
+    private final RestChannel channel;
+    /**
+     * The deprecation message.
+     */
+    private final String warning;
+
+    /**
+     * Create a new {@link DeprecationRestChannel}.
+     *
+     * @param channel The channel to proxy
+     * @param warning The warning to add to the response. This should only be ASCII as it goes into a header!
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestChannel(RestChannel channel, String warning) {
+        assert Strings.validHeaderValue(warning);
+
+        // required
+        this.channel = Objects.requireNonNull(channel);
+        this.warning = Objects.requireNonNull(warning);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Adds the {@link #DEPRECATION_HEADER} to the {@code response}.
+     *
+     * @param response The response to send to the client
+     */
+    @Override
+    public void sendResponse(RestResponse response) {
+        response.addHeader(DEPRECATION_HEADER, warning);
+
+        // actually send the response
+        channel.sendResponse(response);
+    }
+
+    /**
+     * {@inheritDoc}
```

Comment:
> If we're writing and reviewing code carefully
> It's better than we reach an understanding on this.

I disagree with the notion that the reviews make it acceptable to drop it because _only_ the submitter and reviewer(s) are aware of what happened, the reasoning, and any assumptions that goes into the changes. Particularly because of Guice, it's extremely annoying to jump through the code with undocumented-after-undocumented thing, having to dig into the weeds to find things that should be pretty simple. (This is more in regard to our general disregard for Javadoc, not `inheritdoc`)

I rarely document things for myself because I inherently know what I did (I'd hope, right? :)); it's intended for someone down the road that is not coming in with my biases and assumptions, or a total awareness of the entire codebase. In the case of `inheritdoc`, I agree it's _much_ looser than full documentation, but the explicit nature of it is far different as a reader to show that it's not supposed to change from parent behavior, especially in the context of the rest of our codebase. Furthermore, it suggests to any future reviewer to have someone changing the method to document their changes rather than just keep that as-is in a review somewhere. I got this way because I've inherited too many bad codebases (not suggesting that's what we have by any stretch) where they didn't document anything _and_ the code was bad/broken, which led to all sorts of pain whenever anything needed to be changed by a new member of the team. Applying `inheritdoc` also starts a good practice of actually adding Javadoc.

For me, it's very different from using/not using _syntax_ like `this` because that requires you to read the code either way. Documentation is supposed to help to give context about the contract without reading the raw implementation (e.g., side-effects, can it ever return `null`, when does it throw exceptions, etc.). It's like looking up JDK classes without looking up their implementation. For very advanced stuff, I certainly jump around the implementation, but for simple things it's nice to always know about explicit intent. I think our own inherent disagreement on when to use it shows that the implicit nature leads to different assumptions.